lib/src/rules/diagnostic_describe_all_properties.dart

```diff
@@ -0,0 +1,248 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/ast.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'DO reference all public properties in debug methods.';
+
+const _details = r'''
+**DO** reference all public properties in `debug` method implementations.
+
+Implementers of `Diagnosticable` should reference all public properties in
+a `debugFillProperties(...)` or `debugDescribeChildren(...)` method
+implementation to improve debuggability at runtime.
+
+Public properties are defined as fields and getters that are
+
+* not package-private (e.g., prefixed with `_`)
+* not `static` or overriding
+* not themselves `Widget`s or collections of `Widget`s
```

Comment:
maybe explain why widget is special cased from the general motivation that properties that describe parent-child relationships do not need to be in debugFillProperties.